### ISSUE TITLE
Add cookie changes to migration guide

### ISSUE THREAD

**Issue by benmccann** [playframework/playframework#4460]
Play 2.3 and 2.4-RC2 and earlier did not check cookie validity against the cookie RFC. This means that previously you could set values with characters that are no longer allowed. E.g. you could store JSON in a cookie previously and now that will fail without encoding it first. While this is against the RFC, it worked perfectly fine in browsers depending upon which characters and browsers were being used.

When I read a cookie with Play 2.4-RC3 that was set with Play 2.3, the cookie is not read correctly. Calling `ctx().request().cookie(key)` simply returns `null`.

It seems the recommendation is to encode cookie values as a query parameter (e.g. `encodeURIComponent`). It could be nice for Play to do this automatically for the user when calling `setCookie` and we could have a `setCookieRaw` that doesn't do this.

We should also document how to create a filter to migrate user cookies that did not conform to the RFC to values that can be read by Play 2.4.

> RFC6265:
 cookie-value      = *cookie-octet / ( DQUOTE *cookie-octet DQUOTE )
 cookie-octet      = %x21 / %x23-2B / %x2D-3A / %x3C-5B / %x5D-7E
                       ; US-ASCII characters excluding CTLs,
                       ; whitespace DQUOTE, comma, semicolon,
                       ; and backslash

**Comment by richdougherty** [comment]
Do you think it's OK to return `null` or do you think an error should be thrown instead? The trouble with returning `null` is that you might not even realise that there's an error. Throwing an error has its own problems too.

**Comment by benmccann** [comment]
I think returning `null` as it's currently doing is probably better than throwing an error. If the user gets an error on every single page that they visit that's pretty disruptive compared to if they just get logged out or lose some settings or something. It would be nice to log an error or warning though. Right now it's silent as to the fact that it couldn't parse the cookie.

**Comment by benmccann** [comment]
From @jroper:

>I just remembered, I can’t believe I didn’t think of this earlier, but I exposed a config item to control how strict netty is. either way, it’s still stricter than it used to be (that was necessary to fix the security vulnerability), so I’m not sure if it will help you, but try setting play.http.cookies.strict = false

**Comment by jroper** [comment]
Technically it's not a 2.4 migration task, since these changes also went into Play 2.3.9.

**Comment by rmorais** [comment]
Can someone confirm that Play 2.4 follows RFC6265 for cookies values?

According to that RFC, the character ! should be allowed, but if I try to create a cookie with character, play throws an exception

`Caused by: java.lang.IllegalArgumentException: Cookie value contains an invalid char: !`

I need to create my encoding/decoding functions as I didn't find anything available for Scala, but not sure which RDC should I follow.

**Comment by wsargent** [comment]
rmorias can you provide a unit test or a sample Github project?

**Comment by rmorais** [comment]
@wsargent You can check the sample project here [](https://github.com/rmorais/CookieIssue)

**Comment by wsargent** [comment]
So this 

```
def index = Action {
    val cookie = Cookie("TestCookie", "!")
    Ok(views.html.index("Your new application is ready.")).withCookies(cookie)
}
```

causes:

```

! @6pefc9k15 - Internal server error, for (GET) [/] ->

play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[IllegalArgumentException: Cookie value contains an invalid char: !]]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:265) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:191) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.GlobalSettings$class.onError(GlobalSettings.scala:179) [play_2.11-2.4.6.jar:2.4.6]
	at play.api.DefaultGlobal$.onError(GlobalSettings.scala:212) [play_2.11-2.4.6.jar:2.4.6]
	at play.api.http.GlobalSettingsHttpErrorHandler.onServerError(HttpErrorHandler.scala:94) [play_2.11-2.4.6.jar:2.4.6]
	at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$9$$anonfun$apply$1.applyOrElse(PlayDefaultUpstreamHandler.scala:151) [play-netty-server_2.11-2.4.6.jar:2.4.6]
	at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$9$$anonfun$apply$1.applyOrElse(PlayDefaultUpstreamHandler.scala:148) [play-netty-server_2.11-2.4.6.jar:2.4.6]
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36) [scala-library-2.11.6.jar:na]
	at scala.util.Failure$$anonfun$recover$1.apply(Try.scala:215) [scala-library-2.11.6.jar:na]
	at scala.util.Try$.apply(Try.scala:191) [scala-library-2.11.6.jar:na]
	at scala.util.Failure.recover(Try.scala:215) [scala-library-2.11.6.jar:na]
	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:324) [scala-library-2.11.6.jar:na]
	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:324) [scala-library-2.11.6.jar:na]
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32) [scala-library-2.11.6.jar:na]
	at play.api.libs.iteratee.Execution$trampoline$.executeScheduled(Execution.scala:109) [play-iteratees_2.11-2.4.6.jar:2.4.6]
	at play.api.libs.iteratee.Execution$trampoline$.execute(Execution.scala:71) [play-iteratees_2.11-2.4.6.jar:2.4.6]
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:40) [scala-library-2.11.6.jar:na]
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:248) [scala-library-2.11.6.jar:na]
	at scala.concurrent.Promise$class.complete(Promise.scala:55) [scala-library-2.11.6.jar:na]
	at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:153) [scala-library-2.11.6.jar:na]
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:23) [scala-library-2.11.6.jar:na]
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40) [akka-actor_2.11-2.3.13.jar:na]
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397) [akka-actor_2.11-2.3.13.jar:na]
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260) [scala-library-2.11.6.jar:na]
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339) [scala-library-2.11.6.jar:na]
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979) [scala-library-2.11.6.jar:na]
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107) [scala-library-2.11.6.jar:na]
Caused by: java.lang.IllegalArgumentException: Cookie value contains an invalid char: !
	at play.core.netty.utils.CookieEncoder.validateCookie(CookieEncoder.java:47) ~[play-netty-utils-2.4.6.jar:2.4.6]
	at play.core.netty.utils.ServerCookieEncoder.encode(ServerCookieEncoder.java:77) ~[play-netty-utils-2.4.6.jar:2.4.6]
	at play.api.mvc.Cookies$$anonfun$5.apply(Http.scala:848) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.Cookies$$anonfun$5.apply(Http.scala:841) ~[play_2.11-2.4.6.jar:2.4.6]
	at scala.collection.immutable.Stream.map(Stream.scala:418) ~[scala-library-2.11.6.jar:na]
	at play.api.mvc.Cookies$.encodeSetCookieHeader(Http.scala:841) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.Cookies$.mergeSetCookieHeader(Http.scala:936) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.Result.withCookies(Results.scala:139) ~[play_2.11-2.4.6.jar:2.4.6]
	at controllers.Application$$anonfun$index$1.apply(Application.scala:10) ~[classes/:na]
	at controllers.Application$$anonfun$index$1.apply(Application.scala:8) ~[classes/:na]
	at play.api.mvc.ActionBuilder$$anonfun$apply$17.apply(Action.scala:439) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.ActionBuilder$$anonfun$apply$17.apply(Action.scala:439) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.ActionBuilder$$anonfun$apply$16.apply(Action.scala:408) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.ActionBuilder$$anonfun$apply$16.apply(Action.scala:407) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.Action$.invokeBlock(Action.scala:533) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.Action$.invokeBlock(Action.scala:530) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.ActionBuilder$$anon$1.apply(Action.scala:493) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.Action$$anonfun$apply$1$$anonfun$apply$4$$anonfun$apply$5.apply(Action.scala:105) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.Action$$anonfun$apply$1$$anonfun$apply$4$$anonfun$apply$5.apply(Action.scala:105) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.utils.Threads$.withContextClassLoader(Threads.scala:21) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.Action$$anonfun$apply$1$$anonfun$apply$4.apply(Action.scala:104) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.Action$$anonfun$apply$1$$anonfun$apply$4.apply(Action.scala:103) ~[play_2.11-2.4.6.jar:2.4.6]
	at scala.Option.map(Option.scala:146) ~[scala-library-2.11.6.jar:na]
	at play.api.mvc.Action$$anonfun$apply$1.apply(Action.scala:103) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.mvc.Action$$anonfun$apply$1.apply(Action.scala:96) ~[play_2.11-2.4.6.jar:2.4.6]
	at play.api.libs.iteratee.DoneIteratee$$anonfun$mapM$2.apply(Iteratee.scala:741) ~[play-iteratees_2.11-2.4.6.jar:2.4.6]
	at play.api.libs.iteratee.DoneIteratee$$anonfun$mapM$2.apply(Iteratee.scala:741) ~[play-iteratees_2.11-2.4.6.jar:2.4.6]
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24) [scala-library-2.11.6.jar:na]
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24) [scala-library-2.11.6.jar:na]
	... 6 common frames omitted
```

this comes out to https://github.com/playframework/playframework/blob/master/framework/src/play-netty-utils/src/main/java/play/core/netty/utils/CookieUtil.java

```

    // US-ASCII characters excluding CTLs, whitespace, DQUOTE, comma, semicolon, and backslash
    private static BitSet validCookieValueOctets() {
        BitSet bits = new BitSet(8);
        for (int i = 35; i < 127; i++) {
            // US-ASCII characters excluding CTLs (%x00-1F / %x7F)
            bits.set(i);
        }
        bits.set('"', false);  // exclude DQUOTE = %x22
        bits.set(',', false);  // exclude comma = %x2C
        bits.set(';', false);  // exclude semicolon = %x3B
        bits.set('\\', false); // exclude backslash = %x5C
        return bits;
    }
```

which looks like a copy of http://netty.io/4.0/xref/io/netty/handler/codec/http/cookie/CookieUtil.html

However, the cookie value is defined as `cookie-octet` in https://tools.ietf.org/html/rfc6265#section-4.1.1

```
 cookie-value      = *cookie-octet / ( DQUOTE *cookie-octet DQUOTE )
 cookie-octet      = %x21 / %x23-2B / %x2D-3A / %x3C-5B / %x5D-7E
                       ; US-ASCII characters excluding CTLs,
                       ; whitespace DQUOTE, comma, semicolon,
                       ; and backslash
```

which looks more expansive than the previous code.

There is a fairly good explanation here: http://stackoverflow.com/a/1969339/5266 which boils down the cookie value to:

> So that leaves the alphanums plus "!#$%&'()*+-./:<=>?@[]^_`{|}~"

Netty bug is https://github.com/netty/netty/pull/3239

**Comment by wsargent** [comment]
Latest 4.1 is https://github.com/netty/netty/blob/4.1/codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieEncoder.java

**Comment by wsargent** [comment]
This is interesting: Netty 3.10 fails with this

```
"properly encode ! character" in {
      val encoder =  org.jboss.netty.handler.codec.http.cookie.ServerCookieEncoder.STRICT
      val output = encoder.encode("TestCookie", "!")
      output must be_==("TestCookie=!")
    }
```

but works with 

```
    "properly encode everything but ! character" in {
      val encoder =  org.jboss.netty.handler.codec.http.cookie.ServerCookieEncoder.STRICT
      val output = encoder.encode("TestCookie", "#$%&'()*+-./:<=>?@[]^_`{|}~")
      output must be_==("TestCookie=#$%&'()*+-./:<=>?@[]^_`{|}~")
    }
```

**Comment by wsargent** [comment]
Paging @slandelle as this is just plain odd

**Comment by slandelle** [comment]
Explanation is very simple: I messed up and missed 0x33 (!). Will fix.

**Comment by slandelle** [comment]
Released AHC 1.9.36 and 2.0.0-RC14 with the fix.
Sent PR for Netty: https://github.com/netty/netty/pull/4998

**Comment by wsargent** [comment]
Good find @rmorais & @slandelle.  Will add a fix for the Play copied code and backport to 2.4.x...
